### PR TITLE
Basic hip support

### PR DESCRIPTION
**Pull request by mminutoli**
Hello,

I am opening this pull request to start the process of enabling HIP support in your library. These patches were based on a previous release so they might need rework. I am happy to help and or get some guidance from you.

### CODE REVIEW

**Comment by mminutoli** [comment]
I want to apologize I should have arranged this better.  

This pull request also adds options to disable the compilation of tests and the examples.  Some of them use the deprecated C++ binding of MPI that are not shipped on any of the facilities I have access to.

I am aware there is a merge conflict. I'll fix it. Anyway, more than to merge it right away, I more interested in feedback from you and in understanding what else might be needed to support HIP.

**Comment by rabauke** [comment]
Dear Marco,

thanks for your contribution, it is highly appreciated. Though, I hesitate to merge the PR as it is.

- The PR mixes up several unrelated improvements. I prefer to merge them separately.
- The observation of the missing `TRNG_CUDA_ENABLE` macro in [trng/yarn4.hpp](https://github.com/rabauke/trng4/pull/29/files#diff-c211f2f5d1145c8907271a8ee833b6045dd07c8fc953bd91ff2875bbdb7cac9d) is a nice catch. I have fixed this, see https://github.com/rabauke/trng4/commit/0a66fd70482334bc4d6b2a877cc73d4716b3ec74.
- Having an option to switch off the compilation of tests and examples is certainly something useful, though I do not fully understand your motivation. TRNG no longer uses the deprecated C++ MPI API, see https://github.com/rabauke/trng4/commit/5e41ca15acd0414ffccdf8f9ae73168bba059068.
- I am not familiar with HIP. Can you please give me some pointers to the HIP documentation? How do you use TRNG4 with HIP?

File: trng/limits.hpp [hunk 1]
```diff
@@ -38,7 +38,7 @@
 #include <cfloat>
 #include <trng/cuda.hpp>
 
-#if defined TRNG_CUDA
+#if defined TRNG_CUDA && !defined __HIPCC__
```
Comment (rabauke) [inline trng/limits.hpp:41]: In the context of adding HIP support, this should rather be 
```
#if defined __CUDACC__
```

Also the macro `TRNG_CUDA` changes is meaning, it' not longer CUDA related.  I tend to replace it by something else, e.g., `TRNG_DEVICE_ATTRIBUTE`.

**Comment by rabauke** [comment]
Fixed indirectly via b2fedc0fdd9fd1e8c0d290c2cb524bed44f92397.

File: trng/limits.hpp (hunk 1 above)
Comment (mminutoli) [reply trng/limits.hpp:41]: I agree with you. It should have been renamed TRNG_GPU_SUPPORT or something similar but I felt it was a major refactoring. Therefore before undertaking it I wanted to check with you.

**Comment by mminutoli** [comment]
> Fixed indirectly via [b2fedc0](https://github.com/rabauke/trng4/commit/b2fedc0fdd9fd1e8c0d290c2cb524bed44f92397).

Sorry, I missed the party apparently. I'll give the new version a try. Thanks for including HIP support. I have been using your library on Frontier ;-).

**Pull request closed without merging**